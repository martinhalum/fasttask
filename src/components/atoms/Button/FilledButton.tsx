/**
 *
 * Button
 *
 */

import React from 'react';
import {Text, TouchableOpacity, View} from 'react-native';
import Icon from 'react-native-vector-icons/FontAwesome5';

import ButtonStyles from './styles';
import type {PropsType} from './types';
import {MainTheme} from 'theme';

/**
 * Renders a button with customizable styles and functionality.
 *
 * @param label - The text to be displayed on the button.
 * @param leftAlign - Determines if the button's icon should be aligned to the left. Defaults to `false`.
 * @param leftIcon - The name of the icon to be displayed on the left side of the button.
 * @param rightIcon - The name of the icon to be displayed on the right side of the button.
 * @param type - Determines the style of the button. Defaults to 'primary'.
 * @param customColor - Overrides the default background color of the button.
 * @param textColor - Overrides the default text color of the button.
 * @param disabled - Determines if the button is disabled. Defaults to `false`.
 * @param onPress - The function to be called when the button is pressed.
 *
 * @returns The rendered button component.
 */
function FilledButton({
  label,
  leftAlign,
  leftIcon,
  rightIcon,
  type,
  customColor,
  textColor,
  disabled,
  onPress,
}: PropsType): React.ReactElement {
  const backgroundColor =
    leftIcon !== undefined ? MainTheme.colors.primary : 'transparent';

  const containerStyle =
    type === 'primary'
      ? ButtonStyles.primaryContainer
      : ButtonStyles.secondaryContainer;
  return (
    <TouchableOpacity
      role="button"
      style={
        customColor === undefined
          ? containerStyle
          : {...containerStyle, backgroundColor: customColor}
      }
      onPress={onPress}
      disabled={disabled}>
      {!leftAlign && (
        <View
          style={[
            ButtonStyles.iconContainer,
            {backgroundColor: backgroundColor},
          ]}>
          {leftIcon && <Icon name={leftIcon} style={{color: 'white'}} />}
        </View>
      )}
      <Text role="heading" style={{color: textColor}}>
        {label}
      </Text>
      <View style={ButtonStyles.iconContainer}>
        {rightIcon && (
          <Icon name={rightIcon} size={15} style={{color: textColor}} />
        )}
      </View>
    </TouchableOpacity>
  );
}

export default FilledButton;
