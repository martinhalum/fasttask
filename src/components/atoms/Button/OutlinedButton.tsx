/**
 *
 * Button
 *
 */

import React from 'react';
import {Text, TouchableOpacity} from 'react-native';

import ButtonStyles from './styles';

import type {PropsType} from './types';

/**
 * Renders a button with different styles based on the `type` prop.
 *
 * @param label - The text to be displayed on the button.
 * @param type - The style type of the button. It can be one of the following: 'primary', 'secondary', 'outlined', or 'submit'.
 * @param onPress - The callback function to be executed when the button is pressed.
 * @param disabled - Determines if the button is disabled or not.
 *
 * @returns The rendered button component.
 */
function OutlinedButton({
  label,
  type,
  onPress,
  disabled,
}: PropsType): React.ReactElement {
  return (
    <TouchableOpacity
      style={
        type === 'outlined'
          ? ButtonStyles.outlinedContainer
          : ButtonStyles.submitContainer
      }
      onPress={onPress}
      disabled={disabled}>
      <Text
        style={
          type === 'outlined'
            ? ButtonStyles.outlinedLabel
            : ButtonStyles.submitLabel
        }>
        {label}
      </Text>
    </TouchableOpacity>
  );
}

export default OutlinedButton;
