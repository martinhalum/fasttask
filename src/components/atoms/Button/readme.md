# Button

## Summary

The `Button` function is a higher-level component that renders either a `FilledButton` or an `OutlinedButton` based on the `type` prop passed to it.

## Example Usage

```
<Button label="SampleLabel" type="Outlined" onPress={()=>{}}/>
```

### Props

- `label` (string): The text to be displayed on the button.
- `leftAlign` (boolean, optional): Determines if the button should align its content to the left.
- `leftIcon` (string, optional): The name of the icon to be displayed on the left side of the button.
- `rightIcon` (string, optional): The name of the icon to be displayed on the right side of the button.
- `type` ('primary' | 'secondary' | 'outlined' | 'submit', optional): The type of the button.
- `customColor` (string, optional): The custom background color of the button.
- `textColor` (string, optional): The color of the button text.
- `disabled` (boolean, optional): Determines if the button should be disabled.
- `onPress` (() => void, optional): The function to be called when the button is pressed.
