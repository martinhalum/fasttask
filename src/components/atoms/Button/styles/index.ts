/**
 *
 * Button Styles
 *
 */

import {StyleSheet} from 'react-native';
import {MainTheme} from 'theme';

const container = {
  padding: 14,
  borderRadius: 8,
  minHeight: 52,
};

const ButtonStyles = StyleSheet.create({
  primaryContainer: {
    ...container,
    width: '100%',
    justifyContent: 'space-between',
    alignItems: 'center',
    flexDirection: 'row',
    backgroundColor: MainTheme.colors.primary,
  },
  secondaryContainer: {
    ...container,
    width: '100%',
    justifyContent: 'space-between',
    alignItems: 'center',
    flexDirection: 'row',
    backgroundColor: MainTheme.colors.secondary,
  },
  outlinedContainer: {
    ...container,
    width: '100%',
    justifyContent: 'center',
    alignItems: 'center',
    flexDirection: 'row',
    backgroundColor: 'transparent',
    borderWidth: 0.5,
    borderColor: MainTheme.colors.buttonBorder,
  },
  outlinedLabel: {
    color: MainTheme.colors.buttonBorderText,
  },
  submitContainer: {
    ...container,
    width: '100%',
    justifyContent: 'center',
    alignItems: 'center',
    flexDirection: 'row',
    backgroundColor: MainTheme.colors.submitBackground,
  },
  submitLabel: {
    color: MainTheme.colors.submitLabel,
    fontSize: 18,
    fontWeight: '600',
  },
  iconContainer: {
    padding: 6,
    borderRadius: 12,
    height: 25,
    width: 25,
  },
});

export default ButtonStyles;
