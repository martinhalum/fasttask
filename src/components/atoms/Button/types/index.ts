/**
 *
 * Button Types
 *
 */

export type PropsType = {
  label: string;
  leftIcon?: string;
  rightIcon?: string;
  type?: 'primary' | 'secondary' | 'outlined' | 'submit';
  disabled?: boolean;
  onPress?: () => void;
};
