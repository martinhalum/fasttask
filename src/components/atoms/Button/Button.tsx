/**
 *
 * Button
 *
 */

import React from 'react';

import FilledButton from './FilledButton';

import type {PropsType} from './types';
import OutlinedButton from './OutlinedButton';

/**
 * Renders a button based on the provided props.
 *
 * @param label - The text to be displayed on the button.
 * @param leftAlign - Determines if the button's content should be left-aligned. Defaults to `false`.
 * @param leftIcon - The name of the icon to be displayed on the left side of the button.
 * @param rightIcon - The name of the icon to be displayed on the right side of the button.
 * @param type - The type of the button. Defaults to 'primary'.
 * @param customColor - The custom background color of the button.
 * @param textColor - The color of the button's text.
 * @param disabled - Determines if the button should be disabled. Defaults to `false`.
 * @param onPress - The function to be called when the button is pressed.
 * @returns A React element representing a button component based on the provided props.
 */

function Button({
  label,
  leftAlign,
  leftIcon,
  rightIcon,
  type,
  customColor,
  textColor,
  disabled,
  onPress,
}: PropsType): React.ReactElement {
  switch (type) {
    case 'outlined':
    case 'submit':
      return (
        <OutlinedButton
          label={label}
          leftAlign={leftAlign}
          leftIcon={leftIcon}
          rightIcon={rightIcon}
          type={type}
          onPress={onPress}
          customColor={customColor}
          textColor={textColor}
          disabled={disabled}
        />
      );
    default:
      return (
        <FilledButton
          label={label}
          leftAlign={leftAlign}
          leftIcon={leftIcon}
          rightIcon={rightIcon}
          type={type}
          onPress={onPress}
          customColor={customColor}
          textColor={textColor}
          disabled={disabled}
        />
      );
  }
}

export default Button;
