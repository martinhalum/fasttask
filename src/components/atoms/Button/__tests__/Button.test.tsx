/**
 *
 * Button Test
 *
 */

import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';

import Button from '../Button';
import FilledButton from '../FilledButton';

describe('Button', () => {
  it('should render a primary FilledButton with label and onPress function', () => {
    const onPress = jest.fn();
    const tree = render(
      <Button label="Test" type="primary" onPress={onPress} />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render an outlined Button with label and onPress function', () => {
    const onPress = jest.fn();
    const tree = render(
      <Button label="Test" type="outlined" onPress={onPress} />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a submit Button with label and onPress function', () => {
    const onPress = jest.fn();
    const tree = render(
      <Button label="Test" type="submit" onPress={onPress} />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a FilledButton with undefined leftIcon and rightIcon', () => {
    const onPress = jest.fn();
    const tree = render(
      <Button label="Test" type="primary" onPress={onPress} />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a FilledButton with undefined customColor and textColor', () => {
    const onPress = jest.fn();
    const tree = render(
      <Button label="Test" type="primary" onPress={onPress} />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a FilledButton with undefined type', () => {
    const onPress = jest.fn();
    const tree = render(<Button label="Test" onPress={onPress} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('FilledButton', () => {
  it('should render button with label and onPress function', () => {
    const label = 'Submit';
    const onPress = jest.fn();

    const screen = render(<FilledButton label={label} onPress={onPress} />);

    expect(screen.getByTestId('label')).toBeInTheDocument();
    expect(screen.getByTestId('button')).toHaveAttribute('disabled', null);
    fireEvent(screen.getByTestId('button'));
    expect(onPress).toHaveBeenCalled();
  });
});
