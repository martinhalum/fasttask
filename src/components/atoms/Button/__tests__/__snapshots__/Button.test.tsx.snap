// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render a FilledButton with undefined customColor and textColor 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#448AFF",
      "borderRadius": 8,
      "flexDirection": "row",
      "justifyContent": "space-between",
      "minHeight": 52,
      "opacity": 1,
      "padding": 14,
      "width": "100%",
    }
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 15,
          "height": 30,
          "justifyContent": "center",
          "width": 30,
        },
        {
          "backgroundColor": "transparent",
        },
      ]
    }
  />
  <Text
    role="heading"
    style={
      {
        "color": undefined,
      }
    }
  >
    Test
  </Text>
  <View
    style={
      {
        "alignItems": "center",
        "borderRadius": 15,
        "height": 30,
        "justifyContent": "center",
        "width": 30,
      }
    }
  />
</View>
`;

exports[`Button should render a FilledButton with undefined leftIcon and rightIcon 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#448AFF",
      "borderRadius": 8,
      "flexDirection": "row",
      "justifyContent": "space-between",
      "minHeight": 52,
      "opacity": 1,
      "padding": 14,
      "width": "100%",
    }
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 15,
          "height": 30,
          "justifyContent": "center",
          "width": 30,
        },
        {
          "backgroundColor": "transparent",
        },
      ]
    }
  />
  <Text
    role="heading"
    style={
      {
        "color": undefined,
      }
    }
  >
    Test
  </Text>
  <View
    style={
      {
        "alignItems": "center",
        "borderRadius": 15,
        "height": 30,
        "justifyContent": "center",
        "width": 30,
      }
    }
  />
</View>
`;

exports[`Button should render a FilledButton with undefined type 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#D9D9D9",
      "borderRadius": 8,
      "flexDirection": "row",
      "justifyContent": "space-between",
      "minHeight": 52,
      "opacity": 1,
      "padding": 14,
      "width": "100%",
    }
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 15,
          "height": 30,
          "justifyContent": "center",
          "width": 30,
        },
        {
          "backgroundColor": "transparent",
        },
      ]
    }
  />
  <Text
    role="heading"
    style={
      {
        "color": undefined,
      }
    }
  >
    Test
  </Text>
  <View
    style={
      {
        "alignItems": "center",
        "borderRadius": 15,
        "height": 30,
        "justifyContent": "center",
        "width": 30,
      }
    }
  />
</View>
`;

exports[`Button should render a primary FilledButton with label and onPress function 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#448AFF",
      "borderRadius": 8,
      "flexDirection": "row",
      "justifyContent": "space-between",
      "minHeight": 52,
      "opacity": 1,
      "padding": 14,
      "width": "100%",
    }
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 15,
          "height": 30,
          "justifyContent": "center",
          "width": 30,
        },
        {
          "backgroundColor": "transparent",
        },
      ]
    }
  />
  <Text
    role="heading"
    style={
      {
        "color": undefined,
      }
    }
  >
    Test
  </Text>
  <View
    style={
      {
        "alignItems": "center",
        "borderRadius": 15,
        "height": 30,
        "justifyContent": "center",
        "width": 30,
      }
    }
  />
</View>
`;

exports[`Button should render a submit Button with label and onPress function 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#00E676",
      "borderRadius": 8,
      "flexDirection": "row",
      "justifyContent": "center",
      "minHeight": 52,
      "opacity": 1,
      "padding": 14,
      "width": "100%",
    }
  }
>
  <Text
    style={
      {
        "color": "#000000",
        "fontSize": 18,
        "fontWeight": "600",
      }
    }
  >
    Test
  </Text>
</View>
`;

exports[`Button should render an outlined Button with label and onPress function 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "borderColor": "#000000",
      "borderRadius": 8,
      "borderWidth": 0.5,
      "flexDirection": "row",
      "justifyContent": "center",
      "minHeight": 52,
      "opacity": 1,
      "padding": 14,
      "width": "100%",
    }
  }
>
  <Text
    style={
      {
        "color": "#000000",
      }
    }
  >
    Test
  </Text>
</View>
`;
